### PR TITLE
Update TLS certificate structure and secure communication between components

### LINKED ISSUE #1589: Secured Communication (MTLS) between the adapter, router and enforcer
### Describe your problem(s)
The xds server communication (router-adapter and enforcer-adapter) needs to happen based on MTLS. It provides secured communication as well as authentication.
The router calls the enforcer to engage filters. That communication also needs to be based on MTLS.

### How will you implement it

---

### LINKED ISSUE #1596: Have a unified format across all component for TLS certificates
### Describe your problem(s)
It is better to provide a unified way to provide keys and certificates. At the moment, the adapter and the router accept certs in PEM file format while the enforcer needs a JKS file.  The enforcer's certificate structure needs to be changed to PEM format. 
Hence the configuration also may be subjected to change.

### How will you implement it

---

### PR DESCRIPTION
**Pull request by VirajSalaka**
### Purpose
PEM format files need to be provided for all the components.
MTLS based communication among router, enforcer, and adapter.
Provide the capability to assign different labels for the router.

### Issues

Fixes #1596 
Fixes #1589

### Automation tests
 - Unit tests added: No
 - Integration tests added: Yes (Existing test cases covers this aspect)

### Tested environments

MacOS Mojave

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [ ] Assigned the project
- [ ] Validated respective github issues
- [ ] Assigned milestone to the github issue(s)

### CODE REVIEW

File: resources/envoy.yaml [hunk 1]
```diff
@@ -30,10 +30,42 @@ static_resources:
...
+        name: "envoy.transport_sockets.tls"
+        typed_config: 
+          "@type": type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext
+          common_tls_context:
+            tls_params:
+              tls_minimum_protocol_version: TLSv1_2
+              tls_maximum_protocol_version: TLSv1_2
+            tls_certificates:
+              private_key:  
+                filename: "/home/wso2/security/mg.key"
```
Comment (VirajSalaka) [inline resources/envoy.yaml:43]: Check if this can be derived from env variables

File: resources/docker-compose.yaml [hunk 2]
```diff
@@ -11,19 +11,25 @@ services:
...
+    image: wso2/mg-enforcer:4.0.0-m4-SNAPSHOT
     volumes:  
       - ./resources/enforcer/security:/home/wso2/mg/security
       - ./resources/conf:/home/wso2/conf
+    links:
+      - adapter
     environment:
+      - ENFORCER_PRIVATE_KEY_PATH=/home/wso2/mg/security/mg.key
+      - ENFORCER_PUBLIC_KEY_PATH=/home/wso2/mg/security/mg.pem
+      - TRUSTED_ADAPTER_CERTS_PATH=/home/wso2/mg/security
```
Comment (VirajSalaka) [inline resources/docker-compose.yaml:31]: introduce separate folder for truststore and keystore

File: resources/conf/config.toml [hunk 3]
```diff
@@ -55,15 +61,12 @@ disableSslVerification = false
 
 #Key Store Configuration
 [enforcer.keystore]
```
Comment (VirajSalaka) [inline resources/conf/config.toml:63]: Remove this as the truststore is completely picked from env variables.

File: adapter/pkg/mgw/mgw.go [hunk 4]
```diff
@@ -53,6 +57,10 @@ var (
 	alsPort     uint
 
 	mode string
+
```
Comment (VirajSalaka) [inline adapter/pkg/mgw/mgw.go:60]: Remove these variables as unused.

File: adapter/pkg/mgw/mgw.go [hunk 5]
```diff
@@ -73,7 +81,24 @@ const grpcMaxConcurrentStreams = 1000000
 func runManagementServer(server xdsv3.Server, enforcerServer xdsv3.Server, port uint) {
 	var grpcOptions []grpc.ServerOption
 	grpcOptions = append(grpcOptions, grpc.MaxConcurrentStreams(grpcMaxConcurrentStreams))
-	grpcServer := grpc.NewServer()
+
+	cert, err := tlsutils.GetServerCertificate()
+
+	// TODO: (VirajSalaka) add properly
```
Comment (VirajSalaka) [inline adapter/pkg/mgw/mgw.go:87]: Remove todo as it is fixed already.

File: adapter/pkg/mgw/mgw.go [hunk 6]
```diff
@@ -73,7 +81,24 @@ const grpcMaxConcurrentStreams = 1000000
 func runManagementServer(server xdsv3.Server, enforcerServer xdsv3.Server, port uint) {
 	var grpcOptions []grpc.ServerOption
 	grpcOptions = append(grpcOptions, grpc.MaxConcurrentStreams(grpcMaxConcurrentStreams))
-	grpcServer := grpc.NewServer()
+
+	cert, err := tlsutils.GetServerCertificate()
```
Comment (VirajSalaka) [inline adapter/pkg/mgw/mgw.go:85]: Exit the code if error pops

File: adapter/pkg/mgw/mgw.go [hunk 7]
```diff
@@ -73,7 +81,24 @@ const grpcMaxConcurrentStreams = 1000000
...
+
```
Comment (VirajSalaka) [inline adapter/pkg/mgw/mgw.go:86]: Check the behavior when enforcer and router running but adapter killed in the middle.

File: adapter/pkg/synchronizer/apis_fetcher.go [hunk 8]
```diff
@@ -96,6 +96,7 @@ func FetchAPIs(id *string, gwLabel *string, c chan SyncAPIResponse) {
 		if err != nil {
 			logger.LoggerSync.Errorf("Error occurred when readin the cert form %v : %v", certPath, err)
 		}
+		// TODO: (VirajSalaka) Reuse the adapter truststore configuration ?
```
Comment (VirajSalaka) [inline adapter/pkg/synchronizer/apis_fetcher.go:99]: Reuse the existing truststore.

Comment (VirajSalaka) [reply adapter/pkg/synchronizer/apis_fetcher.go:99]: do the same for consul certificate as well.

File: adapter/pkg/tlsutils/tls_utils.go [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+	certContentPattern := `\-\-\-\-\-BEGIN\sCERTIFICATE\-\-\-\-\-((.|\n)*)\-\-\-\-\-END\sCERTIFICATE\-\-\-\-\-`
+	regex := regexp.MustCompile(certContentPattern)
+	if regex.Match(certContent) {
+		return true
+	}
+	return false
+}
+
+// IsPrivateKey checks if the file content represents valid private KEY in PEM format.
+func IsPrivateKey(keyContent []byte) bool {
```
Comment (VirajSalaka) [inline adapter/pkg/tlsutils/tls_utils.go:97]: Remove as this method is unused.

File: adapter/pkg/tlsutils/tls_utils.go [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+// the provided directory/file path.
+func GetTrustedCertPool() *x509.CertPool {
+	onceTrustedCertsRead.Do(func() {
+		caCertPool = x509.NewCertPool()
+		conf, _ := config.ReadConfigs()
+		filepath.Walk(conf.Adapter.Truststore.Location, func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				logger.LoggerTLSUtils.Warn("Error while reading the trusted certificates directory/file.", err)
+			} else {
+				if !info.IsDir() && (filepath.Ext(info.Name()) == ".crt" || filepath.Ext(info.Name()) == ".pem") {
```
Comment (VirajSalaka) [inline adapter/pkg/tlsutils/tls_utils.go:69]: Introduce variables

File: adapter/pkg/xds/xds_server.go [hunk 11]
```diff
@@ -313,15 +314,12 @@ func generateEnforcerConfigs(config *config.Config) *enforcer.Config {
 	}
 
 	return &enforcer.Config{
-		Truststore: &enforcer.CertStore{
```
Comment (VirajSalaka) [inline adapter/pkg/xds/xds_server.go:316]: Remove the code segment as this configuration is no longer used from enforcer.

File: distribution/src/main/assembly/assembly.xml [hunk 12]
```diff
@@ -48,5 +48,29 @@
             <source>../resources/docker-compose.yaml</source>
             <outputDirectory>wso2am-micro-gw-${project.version}</outputDirectory>
         </file>
+        <file>
```
Comment (VirajSalaka) [inline distribution/src/main/assembly/assembly.xml:51]: update k8s artifacts

File: enforcer/src/main/java/org/wso2/micro/gateway/enforcer/config/ConfigHolder.java [hunk 13]
```diff
@@ -147,14 +160,15 @@ private void populateJWTIssuerConfiguration(List<Issuer> cdsIssuers)  {
             issuerDto.setJwksConfigurationDTO(jwksConfigurationDTO);
 
             String certificateAlias = jwtIssuer.getCertificateAlias();
-            try {
-                if (trustStore.getCertificate(certificateAlias) != null) {
-                    Certificate issuerCertificate = trustStore.getCertificate(certificateAlias);
-                    issuerDto.setCertificate(issuerCertificate);
+            if (certificateAlias != null) {
+                // TODO: (VirajSalaka) Decide if it is required to have a alias
```
Comment (VirajSalaka) [inline enforcer/src/main/java/org/wso2/micro/gateway/enforcer/config/ConfigHolder.java:164]: Keep this as for future reference, when the certificate is a payload.

File: enforcer/src/main/java/org/wso2/micro/gateway/enforcer/config/ConfigHolder.java [hunk 14]
```diff
@@ -137,7 +143,14 @@ private void populateEventHub(EventHub eventhub) {
         config.setEventHub(eventHubDto);
     }
 
-    private void populateJWTIssuerConfiguration(List<Issuer> cdsIssuers)  {
+    private void populateJWTIssuerConfiguration(List<Issuer> cdsIssuers) {
+        try {
+            setTrustStoreForJWT(KeyStore.getInstance(KeyStore.getDefaultType()));
+            getTrustStoreForJWT().load(null);
+            // TODO: (VirajSalaka) Should the microgateway panic or should it proceed
```
Comment (VirajSalaka) [inline enforcer/src/main/java/org/wso2/micro/gateway/enforcer/config/ConfigHolder.java:150]: Remove todo

File: enforcer/src/main/java/org/wso2/micro/gateway/enforcer/util/FilterUtils.java [hunk 15]
```diff
@@ -65,23 +64,23 @@ public static String getMaskedToken(String token) {
...
-                publicCert = trustStore.getCertificate(certAlias);
-            }
-        } catch (KeyStoreException e) {
-            String msg = "Error in retrieving public certificate from the trust store with alias : " + certAlias;
-            log.error(msg, e);
-            throw new MGWException(msg, e);
-        }
-        return publicCert;
-    }
+//    public static Certificate getCertificateFromTrustStore(String certAlias) throws MGWException {
```
Comment (VirajSalaka) [inline enforcer/src/main/java/org/wso2/micro/gateway/enforcer/util/FilterUtils.java:67]: Remove unused code here.

File: enforcer/src/main/java/org/wso2/micro/gateway/enforcer/util/TLSUtils.java [hunk 16]
```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (VirajSalaka) [inline enforcer/src/main/java/org/wso2/micro/gateway/enforcer/util/TLSUtils.java:2]: Update the licence header

File: enforcer/src/main/java/org/wso2/micro/gateway/enforcer/util/TLSUtils.java [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+     *
+     * @param filePath Filepath of the corresponding certificate
+     * @return Certificate
+     */
+    public static Certificate getCertificateFromFile(String filePath)
+            throws CertificateException, IOException {
+        return getCertsFromFile(filePath, true).get(0);
+    }
+
+    public static Certificate getCertificateFromString(String certificateContent)
```
Comment (VirajSalaka) [inline enforcer/src/main/java/org/wso2/micro/gateway/enforcer/util/TLSUtils.java:59]: Javadoc missing.

File: enforcer/src/main/java/org/wso2/micro/gateway/enforcer/util/TLSUtils.java [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+        try (InputStream inputStream = new ByteArrayInputStream(certificateContent.getBytes())) {
+            CertificateFactory fact = CertificateFactory.getInstance(X509);
+            return fact.generateCertificate(inputStream);
+        }
+    }
+
+    /**
+     * Add the certificates to the the truststore.
+     *
+     * @param filePath Filepath of the corresponding certificate or directory containing the certificates
```
Comment (VirajSalaka) [inline enforcer/src/main/java/org/wso2/micro/gateway/enforcer/util/TLSUtils.java:71]: param missing.

File: enforcer/src/main/java/org/wso2/micro/gateway/enforcer/util/TLSUtils.java [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+    public static void addCertsToTruststore(KeyStore trustStore, String filePath) throws IOException {
+        if (!Files.exists(Paths.get(filePath))) {
+            log.error("The provided certificates directory/file path does not exist. : " + filePath);
+            return;
+        }
+        if (Files.isDirectory(Paths.get(filePath))) {
+            log.debug("Provided Path is a directory: " + filePath);
+            Files.walk(Paths.get(filePath)).filter(path -> {
+                Path fileName = path.getFileName();
+                return fileName != null && (fileName.toString().endsWith(".crt") ||
```
Comment (VirajSalaka) [inline enforcer/src/main/java/org/wso2/micro/gateway/enforcer/util/TLSUtils.java:83]: Introduce constants.

File: enforcer/src/main/java/org/wso2/micro/gateway/enforcer/util/TLSUtils.java [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+                    log.debug("Certificate Added to the truststore : " + path.toAbsolutePath());
+                } catch (CertificateException | IOException e) {
+                    log.error("Error while adding certificates to the truststore.", e);
+                }
+            });
+        } else {
+            log.debug("Provided Path is a regular File Path : " + filePath);
+            try {
+                List<Certificate> certificateList = getCertsFromFile(filePath, false);
+                certificateList.forEach(certificate -> {
```
Comment (VirajSalaka) [inline enforcer/src/main/java/org/wso2/micro/gateway/enforcer/util/TLSUtils.java:105]: Move the code to a common function as it is shared.

File: enforcer/src/main/proto/wso2 [hunk 21]
```diff
@@ -1 +1 @@
-/Users/praminda/am/envoy/data-plane-api/wso2/
```
Comment (VirajSalaka) [inline enforcer/src/main/proto/wso2:1]: remove this from the PR.

File: adapter/pkg/mgw/mgw.go (hunk 7 above)
Comment (VirajSalaka) [reply adapter/pkg/mgw/mgw.go:86]: Enforcer also goes down but the router remains in the existing state.

File: adapter/pkg/mgw/mgw.go (hunk 4 above)
Comment (VirajSalaka) [reply adapter/pkg/mgw/mgw.go]: fixed with https://github.com/wso2/product-microgateway/pull/1582/commits/b804e40a426ab4a0c8f06cc8c5d68598fbfc8882

File: adapter/pkg/mgw/mgw.go (hunk 5 above)
Comment (VirajSalaka) [reply adapter/pkg/mgw/mgw.go]: fixed with https://github.com/wso2/product-microgateway/pull/1582/commits/b804e40a426ab4a0c8f06cc8c5d68598fbfc8882

File: adapter/pkg/mgw/mgw.go (hunk 6 above)
Comment (VirajSalaka) [reply adapter/pkg/mgw/mgw.go:81]: fixed with https://github.com/wso2/product-microgateway/pull/1582/commits/b804e40a426ab4a0c8f06cc8c5d68598fbfc8882

File: adapter/pkg/tlsutils/tls_utils.go (hunk 9 above)
Comment (VirajSalaka) [reply adapter/pkg/tlsutils/tls_utils.go]: fixed with https://github.com/wso2/product-microgateway/pull/1582/commits/b804e40a426ab4a0c8f06cc8c5d68598fbfc8882

File: adapter/pkg/tlsutils/tls_utils.go (hunk 10 above)
Comment (VirajSalaka) [reply adapter/pkg/tlsutils/tls_utils.go]: fixed with https://github.com/wso2/product-microgateway/pull/1582/commits/b804e40a426ab4a0c8f06cc8c5d68598fbfc8882

File: enforcer/src/main/java/org/wso2/micro/gateway/enforcer/config/ConfigHolder.java (hunk 14 above)
Comment (VirajSalaka) [reply enforcer/src/main/java/org/wso2/micro/gateway/enforcer/config/ConfigHolder.java]: fixed with https://github.com/wso2/product-microgateway/pull/1582/commits/b804e40a426ab4a0c8f06cc8c5d68598fbfc8882

File: enforcer/src/main/java/org/wso2/micro/gateway/enforcer/util/FilterUtils.java (hunk 15 above)
Comment (VirajSalaka) [reply enforcer/src/main/java/org/wso2/micro/gateway/enforcer/util/FilterUtils.java]: fixed with https://github.com/wso2/product-microgateway/pull/1582/commits/b804e40a426ab4a0c8f06cc8c5d68598fbfc8882

File: enforcer/src/main/java/org/wso2/micro/gateway/enforcer/util/TLSUtils.java (hunk 16 above)
Comment (VirajSalaka) [reply enforcer/src/main/java/org/wso2/micro/gateway/enforcer/util/TLSUtils.java]: fixed with https://github.com/wso2/product-microgateway/pull/1582/commits/b804e40a426ab4a0c8f06cc8c5d68598fbfc8882

File: enforcer/src/main/java/org/wso2/micro/gateway/enforcer/util/TLSUtils.java (hunk 17 above)
Comment (VirajSalaka) [reply enforcer/src/main/java/org/wso2/micro/gateway/enforcer/util/TLSUtils.java:67]: fixed with https://github.com/wso2/product-microgateway/pull/1582/commits/b804e40a426ab4a0c8f06cc8c5d68598fbfc8882

File: enforcer/src/main/java/org/wso2/micro/gateway/enforcer/util/TLSUtils.java (hunk 18 above)
Comment (VirajSalaka) [reply enforcer/src/main/java/org/wso2/micro/gateway/enforcer/util/TLSUtils.java]: fixed with https://github.com/wso2/product-microgateway/pull/1582/commits/b804e40a426ab4a0c8f06cc8c5d68598fbfc8882

File: enforcer/src/main/java/org/wso2/micro/gateway/enforcer/util/TLSUtils.java (hunk 19 above)
Comment (VirajSalaka) [reply enforcer/src/main/java/org/wso2/micro/gateway/enforcer/util/TLSUtils.java]: fixed with https://github.com/wso2/product-microgateway/pull/1582/commits/b804e40a426ab4a0c8f06cc8c5d68598fbfc8882

File: enforcer/src/main/java/org/wso2/micro/gateway/enforcer/util/TLSUtils.java (hunk 20 above)
Comment (VirajSalaka) [reply enforcer/src/main/java/org/wso2/micro/gateway/enforcer/util/TLSUtils.java]: fixed with https://github.com/wso2/product-microgateway/pull/1582/commits/b804e40a426ab4a0c8f06cc8c5d68598fbfc8882

File: enforcer/src/main/java/org/wso2/micro/gateway/enforcer/common/ReferenceHolder.java [hunk 22]
```diff
@@ -33,10 +32,8 @@
 
     private final Map<String, KeyValidator> keyValidationHandlerMap = new ConcurrentHashMap<>();
     private KeyManagerDataService keyManagerDataService;
-    private final String nodeLabel;
 
     private ReferenceHolder() {
-        nodeLabel = System.getenv(Constants.ENFORCER_LABEL);
```
[the remark this answers is not in the document]
Comment (VirajSalaka) [reply enforcer/src/main/java/org/wso2/micro/gateway/enforcer/common/ReferenceHolder.java:39]: Now it stored within the ConfigHolder
From now on, we do not read env variables straightaway using system.env
It should be taken from envVarConfig object

File: resources/envoy.yaml (hunk 1 above)
Comment (VirajSalaka) [reply resources/envoy.yaml]: fixed with https://github.com/wso2/product-microgateway/pull/1582/commits/391606c1e87d9aad9da15f06caffd3a309a468ac

File: resources/docker-compose.yaml (hunk 2 above)
Comment (VirajSalaka) [reply resources/docker-compose.yaml]: fixed with https://github.com/wso2/product-microgateway/pull/1582/commits/925259a8d10006fb5f6a5c57750be3a910e8410d

File: adapter/pkg/synchronizer/apis_fetcher.go (hunk 8 above)
Comment (VirajSalaka) [reply adapter/pkg/synchronizer/apis_fetcher.go]: fixed with https://github.com/wso2/product-microgateway/pull/1582/commits/02d61d2a72f2d3d089f049ad719122050faab739

File: wso2 [hunk 23]
```diff
@@ -0,0 +1 @@
+/Users/viraj/go/src/github.com/VirajSalaka/data-plane-api/wso2
```
Comment (Rajith90) [inline wso2:1]: What is this file

Comment (VirajSalaka) [reply wso2]: symlink created by mistake
fixed via https://github.com/wso2/product-microgateway/pull/1582/commits/14ae7b18ab1203289ee04a49cbda4d7aa5bfdfa6

**Review by AmaliMatharaarachchi — APPROVED** [review]

**Pull request merged**